### PR TITLE
Containerized Project Integration

### PR DESCRIPTION
**Pull request by alessandrocarminati**
Greetings! 
This to present this pull request that brings enhancements to our project with multiple applications. 
These updates promise to boost productivity and streamline deployment for a smoother development experience. 

Let's take a closer look at the key improvements:

1. Dockerfile Addition: We're introducing the awaited Dockerfile, enabling seamless containerization for our project. With this addition, we bid farewell to deployment challenges and usher in a more straightforward build process.

2. SQL Organization: To enhance database setup, we've divided SQL initial creation statements into two parts. This reorganization ensures a smoother, more manageable database configuration.

3. Refined Database Defaults: In alignment with container requirements, we've made adjustments to the database defaults. The application will now operate seamlessly within the container environment.

4. Empowering Navweb: We've implemented the navweb landing page and fetch page functionalities, elevating the user experience with a polished interface.

5. Streamlined Build: Essential changes have been made to the makefiles to ensure a successful container build. Building the project has become more efficient and hassle-free.

6. go-bind Efficiency: Embracing the efficiency of go-bind, we've reduced file transfers, leading to improved performance and faster execution.

7. ELISA Tell tale Use Case: A default sample database tailored for the ELISA Tell tale use case is now included, simplifying exploration and testing.

8.  Temporary avoidance what appears to be an alpine distribution bug

### CODE REVIEW

**Comment by alessandrocarminati** [comment]
This pull request is meant to solve #1

File: container/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+RUN apk add --no-cache postgresql postgresql-contrib postgresql-client
+RUN apk add --no-cache binutils
+RUN apk add --no-cache graphviz
+
+# setup resources
+
+WORKDIR /app
+COPY postgres_conf_template.conf /tmp
+# pg.db.tar.gz is a link, please set it to empty database, or whatever db you have.
+copy pg.db.tar.gz /tmp
```
Comment (tinez) [inline container/Dockerfile:61]: Let's follow the convention and use capital `COPY` here

File: container/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,111 @@
+# BUILDER
+FROM alpine:latest AS builder
```
Comment (tinez) [inline container/Dockerfile:2]: While I understand the benefits of having a distro with a small footprint as a base, alpine is known to have its set of issues given it's based on musl and not glibc (I can see you already posted some workaround for dot). Recent [news about most prolific package manager leaving alpine](https://www.phoronix.com/news/Alpine-Linux-Prolific-Packager) on top of that make me wonder if we shouldn't be aiming at something else here (fedora/rhel ubi).

File: container/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+# install build dependecies
+RUN apk add --no-cache autoconf automake build-base git
+RUN apk add --no-cache go
+RUN apk add --no-cache linux-headers
+
+#build app
+WORKDIR /build
+
+# Build radare2
+RUN git clone ${RADARE2_GIT_REPO} /build/radare2 && cd /build/radare2 && git checkout ecd8d00aad289abf8f850b45d5f6c2e6b0c15c0f
```
Comment (tinez) [inline container/Dockerfile:22]: It would be good to have an `ARG` for the revision. Maybe also some short explanation why we don't use an official release.

File: container/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+ARG _POSTGRES_IDENT_FILE=pg_ident.conf
+ENV POSTGRES_IDENT_FILE=$_POSTGRES_IDENT_FILE
+ARG _POSTGRES_CONF_FILE=${POSTGRES_DATA_DIR}/${POSTGRES_NAME}/postgresql.conf
+ENV POSTGRES_CONF_FILE=$_POSTGRES_CONF_FILE
+
+ARG _START_SCRIPT_NAME=/usr/bin/start.sh
+ENV START_SCRIPT_NAME=$_START_SCRIPT_NAME
+
+# install dependencies
+RUN apk add --no-cache postgresql postgresql-contrib postgresql-client
```
Comment (tinez) [inline container/Dockerfile:52]: Postgres is a very big gun - was sqlite also considered a candidate?

File: container/postgres_conf_template.conf [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+port = 5432
+max_connections = 100
+unix_socket_directories = '/tmp'
+ssl = off
+shared_buffers = 128MB
+dynamic_shared_memory_type = posix
+max_wal_size = 1GB
+min_wal_size = 80MB
+log_line_prefix = '%m [%p] %q%u@%d '
+log_timezone = 'Europe/Rome'
```
Comment (tinez) [inline container/postgres_conf_template.conf:15]: While Rome is undoubtedly a very beautiful city, it would be better to stick with UTC here and below ;)

File: nav/config/config.go [hunk 6]
```diff
@@ -174,3 +177,16 @@ func validateType(t *string) error {
...
+func validateGType(t *c.OutIMode) error {
+        switch *t {
+        case 0:
+                *t = c.DefaultGOutputType
+                fmt.Printf("No output format specified. Defaulting to %d.\n", c.DefaultGOutputType)
+                return nil
+        case c.OText, c.OPNG, c.OJPG, c.OSVG:
+                return nil
+        default:
+                return fmt.Errorf("invalid graphviz output type: %s\nChoose one of the following: 0, 1, 2, 3", *t)
```
Comment (tinez) [inline nav/config/config.go:190]: It would be nice to include what those integer options mean in the message, i.e. `...following: 0 (text), 1 (png)...`. Also, the suggested values (0-3) are inconsistent with the one defined below (1-4).

File: navweb/stocazzo.sh [hunk 7]
```diff
@@ -0,0 +1,7 @@
+#!/bin/sh
```
Comment (tinez) [inline navweb/stocazzo.sh:1]: I'm very curious what this script is responsible for :)

**Comment by tinez** [comment]
Tried building locally, but I got:
```
make: *** No rule to make target 'install'.  Stop.
Error: building at STEP "RUN cd /build/ksnav/kern_bin_db && make && make install": while running runtime: exit status 2
```
I'm assuming this is because the `install` target is introduced only in this PR?

File: container/Dockerfile (hunk 1 above)
Comment (alessandrocarminati) [reply container/Dockerfile:61]: Sure, thanks

File: container/Dockerfile (hunk 2 above)
Comment (alessandrocarminati) [reply container/Dockerfile:2]: I don't dislike the musl-busybox duo; in fact, it's the driving force behind my choice. Their lightweight nature is the main motivation for my decision.

Initially, my goal was to create a container that could be completely independent of any distribution. Since my application is written in Golang that produces by default static executables, ks-nav itself could operate without any additional dependencies. However, there are dependencies that are unavoidable.

Let's examine these dependencies:

* PostgreSQL is a significant one. For this ks-nav initial release, I prefer to stick with it due to its extensive testing and superior performance in both database construction and diagram creation.

* Currently, Radare2 is used for extracting symbols and cross-references. This task could potentially be achieved with other lightweight solutions, possibly in house made using golang. Over time, I might consider completely removing Radare2 from the loop.

* The introduction of the graphviz-dot dependency was due to a minor bug in the Golang implementation when building the `svg` format. Eventually, I decided to drop this dependency, as evidenced by the workaround implemented.

In summary, I am inclined to remain with Alpine for the time being. Eventually, I plan to transition to a distribution-free container once I've resolved all the issues preventing its immediate implementation.

File: container/Dockerfile (hunk 4 above)
Comment (alessandrocarminati) [reply container/Dockerfile:52]: SQLite3 holds significant potential. 
Firstly, it can be compiled statically, marking a substantial stride towards my original goal of achieving a distribution-free container. 
Its simpler configuration greatly streamlines the mechanics within the container. 
Moreover, its lightweight nature is another advantage. 
Regrettably, based on the tests conducted thus far (albeit not extensively), I have observed a decline in performance for tasks such as creating new databases and utilizing them to generate diagrams.

Before I can consider replacing PostgreSQL with SQLite3, I must identify a potential solution, if one exists, for this performance degradation. 
This entails scheduling a more comprehensive testing session.

In practical terms, two primary factors have deterred me from adopting SQLite3 in the current stage:
1. **Limited Testing**: SQLite3 has received minimal testing within this application. Its compatibility and performance are only marginally confirmed, in contrast to the more robustly validated PostgreSQL.
2. **Performance Discrepancy**: Notably, SQLite3 exhibits noticeably slower performance when compared to PostgreSQL, posing a hurdle to its immediate adoption.

With these challenges in mind, further exploration and analysis are required before SQLite3 can be considered a viable replacement for PostgreSQL.

File: container/postgres_conf_template.conf (hunk 5 above)
Comment (alessandrocarminati) [reply container/postgres_conf_template.conf:15]: I must admit, I selected the timezone without giving it much thought. Is there truly any alternative timezone available for CE(S)T, apart from that one? ;-P

By the way, I understand your perspective, and it does indeed make sense. I will adjust it as per your suggestion.

File: nav/config/config.go (hunk 6 above)
Comment (alessandrocarminati) [reply nav/config/config.go:190]: This issue represents a regression that occurred when the configuration pull request was merged. The inconsistency can be attributed to historical reasons. In any case, it is something I failed to notice and must address. Your keen observation is much appreciated. Thank you for catching this.

File: navweb/stocazzo.sh (hunk 7 above)
Comment (alessandrocarminati) [reply navweb/stocazzo.sh:1]: Oh, this individual should not be present here. It's a remnant from the initial test I conducted to enable WebSocket functionality in the integration with kern_bin_db. I certainly need to remove it. Thank you.

File: container/Dockerfile (hunk 3 above)
Comment (alessandrocarminati) [reply container/Dockerfile:22]: Got it, valuable suggestion. Thanks.

**Comment by alessandrocarminati** [comment]
> Tried building the container locally, but I got:
> 
> ```
> make: *** No rule to make target 'install'.  Stop.
> Error: building at STEP "RUN cd /build/ksnav/kern_bin_db && make && make install": while running runtime: exit status 2
> ```
> 
> I'm assuming this is because the `install` target is introduced only in this PR?

Indeed, in this pull request, I introduced the `install` target to all the `makefile`s. My assumption is that the testing process would involve fetching this branch and temporarily altering the git reference to my forked repository (for the sake of simplicity).

Here's a suggestion, though feel free to employ a different, more convenient approach if you prefer.
```
ARG _KS_NAV_GIT_REPO="https://github.com/alessandrocarminati/ks-nav.git"
ENV KS_NAV_GIT_REPO=$_KS_NAV_GIT_REPO
ARG _KS_NAV_GIT_REPO_BRANCH=ac-container-merge-review
ENV KS_NAV_GIT_REPO_BRANCH=$_KS_NAV_GIT_REPO_BRANCH
```

File: container/Dockerfile (hunk 2 above)
Comment (tinez) [reply container/Dockerfile:2]: Ack, thanks for the thorough explanation on alpine vs others and postgres vs sqlite.
